# <center>Pandas Datetime [25 exercises with solution]</center>

In [2]:
# Importing necessary libraries:

import numpy as np
import pandas as pd

In [3]:
# Knowing about the csv file:

ufo_data = pd.read_csv('ufo.csv', index_col = 'Date_time', parse_dates = True)
ufo_data.head()

city state/province country UFO_shape  \
Date_time                                                                    
1910-06-01 15:00:00           wills point             tx      us     cigar   
1920-06-11 21:00:00                cicero             in      us   unknown   
1929-07-05 14:00:00  buchanan  (or burns)             or      us      disk   
1931-06-01 13:00:00               abilene             ks      us      disk   
1939-06-01 20:00:00              waterloo             al      us  fireball   

                     length_of_encounter_seconds  \
Date_time                                          
1910-06-01 15:00:00                          120   
1920-06-11 21:00:00                           60   
1929-07-05 14:00:00                           60   
1931-06-01 13:00:00                         1800   
1939-06-01 20:00:00                          300   

                    described_duration_of_encounter  \
Date_time                                             
1910-06-01 15:00:00                       2 minutes   
1920-06-11 21:00:00                        1 minute   
1929-07-05 14:00:00                            1min   
1931-06-01 13:00:00                 30 min. approx.   
1939-06-01 20:00:00                       5 minutes   

                                                           description  \
Date_time                                                                
1910-06-01 15:00:00       Cigar shaped object moving from West to East   
1920-06-11 21:00:00  ((NUFORC Note:  Probable hoax.  Note date.  PD...   
1929-07-05 14:00:00  we were traveling east of burns&#44clmbing up ...   
1931-06-01 13:00:00  This is a what my mother related to me regardi...   
1939-06-01 20:00:00  4 high school students see bright&#44 ground-l...   

                    date_documented   latitude   longitude  
Date_time                                                   
1910-06-01 15:00:00      04/16/2005  32.709167  -96.008056  
1920-06-11 21:00:00      05/12/2009  40.123889  -86.013333  
1929-07-05 14:00:00      08/16/2002  43.642500 -118.627500  
1931-06-01 13:00:00      02/14/2006  38.917222  -97.213611  
1939-06-01 20:00:00      02/21/2014  34.918056  -88.064167

![UFO](https://www.w3resource.com/w3r_images/ufo.png)

![Column Name Description](https://i.imgur.com/UrE9JKv.png)

## 25 Exercises Solutions

### 1. Write a Pandas program to create the todays date.

In [4]:
from datetime import date

now = pd.to_datetime(date.today())    # '2019-11-02 00:00:00'
date = str(now).split()[0]            # After split: ['2019-11-02', '00:00:00']
print('Today\'s Date->', date)

Today's Date-> 2021-02-09


### 2. Write a Pandas program to calculate all the sighting days of the unidentified flying object (ufo) from current date.

In [5]:
date_diff = (now - ufo_data.index).days    # difference of corresponding time in days
result = date_diff.to_frame(index = False, name='Diff in Days')
print(result)

     Diff in Days
0           40430
1           36767
2           33456
3           32760
4           29838
..            ...
342          6153
343          5969
344          5937
345          5906
346          6208

[347 rows x 1 columns]


### 3. Write a Pandas program to get the current date, oldest date and number of days between Current date and oldest date of Ufo dataset.

### 4. Write a Pandas program to get all the sighting days of the unidentified flying object (ufo) which are less than or equal to 40 years (365*40 days).

### 5. Write a Pandas program to get all the sighting days of the unidentified flying object (ufo) between 1950-10-10 and 1960-10-10.

### 6. Write a Pandas program to get all the sighting years of the unidentified flying object (ufo) and create the year as column.

### 7. Write a Pandas program to create a plot to present the number of unidentified flying object (UFO) reports per year.

### 8. Write a Pandas program to extract year, month, day, hour, minute, second and weekday from unidentified flying object (UFO) reporting date.

### 9. Write a Pandas program to convert given datetime to timestamp. 

### 10. Write a Pandas program to count year-country wise frequency of reporting dates of unidentified flying object(UFO). 

### 11. Write a Pandas program to extract unique reporting dates of unidentified flying object (UFO).

### 12. Write a Pandas program to get the difference (in days) between documented date and reporting date of unidentified flying object (UFO).

### 13. Write a Pandas program to add 100 days with reporting date of unidentified flying object (UFO). 

### 14. Write a Pandas program to generate sequences of fixed-frequency dates and time spans.

### 15. Write a Pandas program to create a conversion between strings and datetime.

### 16. Write a Pandas program to manipulate and convert date times with timezone information.

### 17. Write a Pandas program to get the average mean of the UFO (unidentified flying object) sighting was reported.

### 18. Write a Pandas program to create a graphical analysis of UFO (unidentified flying object) Sightings year.

### 19. Write a Pandas program to check the empty values of UFO (unidentified flying object) Dataframe.

### 20. Write a Pandas program to create a plot of distribution of UFO (unidentified flying object) observation time.

### 21. Write a Pandas program to create a graphical analysis of UFO (unidentified flying object) sighted by month.

### 22. Write a Pandas program to create a comparison of the top 10 years in which the UFO was sighted vs the hours of the day.

### 23. Write a Pandas program to create a comparison of the top 10 years in which the UFO was sighted vs each Month.

### 24. Write a Pandas program to create a heatmap (rectangular data as a color-encoded matrix) for comparison of the top 10 years in which the UFO was sighted vs each Month. 

### 25. Write a Pandas program to create a Timewheel of Hour Vs Year comparison of the top 10 years in which the UFO was sighted.

Data Sources: ufo_sighting_data.csv - [80,332 records] <br>
Exercises data sources: ufo.csv - [347 random records from ufo_sighting_data.csv] <br>

Note there are no missing data in the columns (ufo.csv).

Source:
- https://github.com/planetsig/ufo-reports 
- https://www.kaggle.com/camnugent/ufo-sightings-around-the-world

#### Reference: [https://www.w3resource.com/python-exercises/pandas/datetime/index.php](https://www.w3resource.com/python-exercises/pandas/datetime/index.php)